File: solutions/haskell/perfect-numbers/1/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify x
    | x <= 0 = Nothing
    | x == 1 = Just Deficient
    | aliquotSum x == x = Just Perfect
    | aliquotSum x < x = Just Deficient
    | aliquotSum x > x = Just Abundant

aliquotSum :: Int -> Int
aliquotSum n = sum [i | i <- [1..(div n 2)], mod n i == 0]

File: solutions/python/perfect-numbers/1/perfect_numbers.py
def classify(number):
    """ A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """
    if number < 1:
        raise ValueError("Classification is only possible for positive integers.")
    nums = []
    for i in range(1, number):
        if number % i == 0:
            nums.append(i)
    s = sum(nums)
    if s == number:
        return "perfect"
    elif s < number:
        return "deficient"
    return "abundant"